Feature/adding logged in support

phyzical opened the pull request:
Creating this pr for initial eyes over the code,

current summary 

* added logic to try next download if 1 fails up to the top 10 links found
* added support for saved albums
* added support for saved playlists
* added support for saved tracks
* various simplifications and refactors
* added logic to always save a track in artist/album/song.mp3
* it will now auto login when a username and password is provided, this is required for non tty due to needing to auth, this is only needed when provided the --sa --st --sp flags due to needing user context. if username and password is not provided a browser is launched for manual user entry 
* tokens only last an hour so i don't think there's privacy concerns
* added support to save all tracks in the outputdir irrelevant of the folder/playlist/album contexts
* added support for users to override the config.js file to change the defaults to avoid adding to commands
* added support for auto removal of sponsors within videos

fixes #90 
fixes #91 
fixes #92 
fixes #93 
fixes #96 
fixes #101 
fixes #103

phyzical commented:
hey @SwapnilSoni1999 when you get a chance mind having a quick glace at the current code to see if there is anything that bothers you while i work on the last few todos

phyzical commented:
note: to test this you will need to add `http://localhost:7654/callback` as an allowed url in the app settings on the spotify developer account

phyzical commented:
@SwapnilSoni1999 this now fixes/implements links issues let me know what you think when you get a chance :) no rush

ajayyy commented on lib/downloader.js:
```diff
@@ -1,20 +1,45 @@
+  let segments = [];
+  let complexFilter = null;
+  try {
+    segments = await sponsorBlock.getSegments(videoID, 'sponsor');
```
If this is for music videos, you might want to put all the categories here, or at least `music_offtopic` which is for non-music sections in music videos.

phyzical replied:
ah good suggestion 👍 (great service btw @pavanjadhaw ) 

question while we have you are all the segments user entered or do you have some "ai/autoscraped" sections too.

i only ask as when it was first requested i was a bit skeptical that all videos would get hits, but when i tested i was impressed with every channel i could think of coming back with hits (though this could just be because they were still somewhat popular channels)

ajayyy replied:
No, it's all manual submissions, but as you noticed, coverage is extremely high, at least in English content released after the the extension started (mid 2019 for sponsor, mid 2020 for other categories).

SwapnilSoni1999 replied:
amazing

SwapnilSoni1999 commented:
I'll check this out in couple of days im sorry for being busy 
Also can you mention a small changelog here in comments please :)

phyzical commented:
@SwapnilSoni1999 No stress dude, life comes first.

the original ticket should reflect all the issues this currently fixes but,

* added logic retry youtube links up to a max of 10 links if there is a failure for whatever reason
* added support for saved albums - requires auth, new flag is --sa
* added support for saved playlists - requires auth, new flag is --sp
* added support for saved tracks - requires auth,  new flag is --st
* Added a new flag --oo which will force all downloaded songs into the output dir (-o flag), without the --oo songs will always save a track in artist/album/song.mp3 irrelevant of what is being downloaded
* Added support for auth of a user to get saved albums, playlists and tracks, If you provide a username and password (--u and --p) we will attempt to autologin, otherwise a  browser prompt will be served for users. when non tty an error will be thrown when attempting --sa,sp or st as we cannot serve a browser without user interaction
* There is now basic support for users to overwrite default flags in config.js 
* added support for auto removal of sponsors and other irrelevant non music sections within songs
* bunch of DRY throughout the code so we end up with the same song data at the download step

SwapnilSoni1999 commented:
Amazing work. I wanted to ask that can you add example commands in README? So it will be useful for all users :)

SwapnilSoni1999 commented:
I've added the callback url to the application settings. I've also tested and its working fine :) great work man, Really appreciated your hard work.
